### ISSUE TITLE
API to serialize and deserialize an NPC

### ISSUE THREAD

**Issue by tastybento** [Pyrbu/ZNPCsPlus#163]
Hi,

**TL;DR** Could you add a way to serialize and deserialize an NPC to the API? Or at least make the serialization and deserialization classes available in the API, please?

I'm the author of BentoBox and I have been asked by admins for them to be able to include ZNPCsPlus NPC's in the starter islands for SkyBlock, OneBlock, AcidIsland, etc. so I am writing a hook that will enable that if ZNPCsPlus is installed.

The starter islands are stored in a blueprint, which is like a schematic. Admins will build the starter island and place various NPCs on it. When the island is copied and stored to file, I serialize the NPCs (and the blocks and entities, etc) and save them. Then when they are pasted they are created from the serialize data and placed at the same relative position to the origin of the island. 

Ideally, it would be great if you could provide a simple call to serialize an NPC, e.g., to YAML and also deserialize YAML into an NPC object. However, alternatively, from your code, I see how you store NPCs using YAML, etc.(e.g. `plugin/src/main/java/lol/pyr/znpcsplus/storage/yaml/YamlStorage.java`) and I'd like to be able to tap into that ability. For example, here's some code that I have been working on:

```
    public String serializeNPC(Npc npc, Vector origin) {
        if (npc == null) {
            throw new IllegalArgumentException("NPC cannot be null.");
        }
        YamlConfiguration config = new YamlConfiguration();
        NpcType type = npc.getType();
        for (EntityProperty<?> property : npc.getAppliedProperties())
            try {
                PropertySerializer<?> serializer = propertyRegistry
                        .getSerializer(((EntityPropertyImpl<?>) property).getType());
                if (serializer == null) {
                    BentoBox.getInstance().logWarning("Unknown serializer for property '" + property.getName()
                            + "' for npc '" + npc.getUuid() + "'. skipping ...");
                    continue;
            }
                config.set("properties." + property.getName(), serializer.UNSAFE_serialize(npc.getProperty(property)));
            } catch (Exception exception) {
                BentoBox.getInstance().logWarning(
                        "Failed to serialize property " + property.getName() + " for npc with id " + npc.getUuid());
                exception.printStackTrace();
        }

        lol.pyr.znpcsplus.api.hologram.Hologram hologram = npc.getHologram();
        if (hologram.getRefreshDelay() != -1)
            config.set("hologram.refresh-delay", hologram.getRefreshDelay());
        List<String> lines = new ArrayList<>(npc.getHologram().lineCount());
        for (int i = 0; i < npc.getHologram().lineCount(); i++) {
            lines.add(hologram.getLine(i));
        }
        config.set("hologram.lines", lines);
        config.set("actions", npc.getActions().stream().map(InteractionAction::toString).filter(Objects::nonNull)
                .collect(Collectors.toList()));
        return config.saveToString();
    }
```

The issue here is that the `PropertySerializer` and `EntityPropertyImpl` classes are not part of the API. I could try and write my own, or indeed copy your serialization code verbatim, but it'd be a lot more efficient and robust if I could use your classes directly. 

Could you make it available? Or is there another option?

**Comment by Pyrbu** [comment]
Does c780c0ec78c8ea422cecbe9da9d16e52c46a368d do everything you wanted?

**Comment by tastybento** [comment]
@Pyrbu Thanks! It looks promising. Is there a snapshot version of the plugin that I can test with? Obviously the current 2.0.0 doesn't have this new API in there.

Below is the code I have using this API for the serialization. The only part I'm not sure about is `config.set("properties." + property.getName(), serializer.serialize(npcEntry));` where the `serialize` method takes the `npcEntry`. Is that correct?

```
    public String serializeNPC(NpcEntry npcEntry, Vector origin) {
        if (npcEntry == null) {
            throw new IllegalArgumentException("NPC cannot be null.");
        }
        YamlConfiguration config = new YamlConfiguration();

        for (EntityProperty<?> property : npcEntry.getNpc().getAppliedProperties())
            try {
                NpcSerializer<?> serializer = reg.getSerializer(property.getClass());
                if (serializer == null) {
                    BentoBox.getInstance().logWarning("Unknown serializer for property '" + property.getName()
                            + "' for npc '" + npcEntry.getId() + "'. skipping ...");
                    continue;
                }
                config.set("properties." + property.getName(), serializer.serialize(npcEntry)); // No need to give the property?
            } catch (Exception exception) {
                BentoBox.getInstance().logWarning(
                        "Failed to serialize property " + property.getName() + " for npc with id " + npcEntry.getId());
                exception.printStackTrace();
            }

        lol.pyr.znpcsplus.api.hologram.Hologram hologram = npcEntry.getNpc().getHologram();
        if (hologram.getRefreshDelay() != -1)
            config.set("hologram.refresh-delay", hologram.getRefreshDelay());
        List<String> lines = new ArrayList<>(npcEntry.getNpc().getHologram().lineCount());
        for (int i = 0; i < npcEntry.getNpc().getHologram().lineCount(); i++) {
            lines.add(hologram.getLine(i));
        }
        config.set("hologram.lines", lines);
        config.set("actions",
                npcEntry.getNpc().getActions().stream().map(InteractionAction::toString).filter(Objects::nonNull)
                .collect(Collectors.toList()));
        String out = config.saveToString();
        BentoBox.getInstance().logDebug(out);
        return out;

    }
```

**Comment by Pyrbu** [comment]
This repository https://repo.pyr.lol/#/snapshots gets updated whenever the ci builds so you can just use that to test, or if you prefer you can clone the repo and publish it to maven local for testing. Also let me know when you're planning to release a new version of your plugin so I can plan the release of the version with the new API accordingly.

**Comment by Pyrbu** [comment]
Also I'm confused about the code you sent, with the new commit you should be able to just do this
```java
return NpcApiProvider.get().getNpcSerializerRegistry().getSerializer(YamlConfiguration.class).serialize(entry).saveToString();
```

**Comment by tastybento** [comment]
Aha! That makes a lot more sense! It wasn't clear what the parameter should be for the serializer but now it is. I'll try this soon.

**Comment by tastybento** [comment]
> This repository https://repo.pyr.lol/#/snapshots gets updated whenever the ci builds so you can just use that to test, or if you prefer you can clone the repo and publish it to maven local for testing. Also let me know when you're planning to release a new version of your plugin so I can plan the release of the version with the new API accordingly.

I tried to use `znpcsplus-api-2.0.0-20241224.151437-145.jar` but I get this error:

```
[10:08:29 ERROR]: [DirectoryProviderSource] Error loading plugin: java.lang.IllegalArgumentException: Directory 'plugins/.paper-remapped/znpcsplus-api-2.0.0-20241224.151437-145.jar' does not contain a paper-plugin.yml or plugin.yml! Could not determine plugin type, cannot load a plugin from it!
java.lang.RuntimeException: java.lang.IllegalArgumentException: Directory 'plugins/.paper-remapped/znpcsplus-api-2.0.0-20241224.151437-145.jar' does not contain a paper-plugin.yml or plugin.yml! Could not determine plugin type, cannot load a plugin from it!
	at io.papermc.paper.plugin.provider.source.FileProviderSource.registerProviders(FileProviderSource.java:85) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at io.papermc.paper.plugin.provider.source.DirectoryProviderSource.registerProviders(DirectoryProviderSource.java:52) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at io.papermc.paper.plugin.provider.source.DirectoryProviderSource.registerProviders(DirectoryProviderSource.java:17) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at io.papermc.paper.plugin.util.EntrypointUtil.registerProvidersFromSource(EntrypointUtil.java:15) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at io.papermc.paper.plugin.PluginInitializerManager.load(PluginInitializerManager.java:113) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at net.minecraft.server.Main.main(Main.java:111) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at io.papermc.paper.PaperBootstrap.boot(PaperBootstrap.java:21) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at org.bukkit.craftbukkit.Main.main(Main.java:281) ~[paper-1.21.4.jar:1.21.4-30-aa2c52b]
	at io.papermc.paperclip.Paperclip.lambda$main$0(Paperclip.java:42) ~[app:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.IllegalArgumentException: Directory 'plugins/.paper-remapped/znpcsplus-api-2.0.0-20241224.151437-145.jar' does not contain a paper-plugin.yml or plugin.yml! Could not determine plugin type, cannot load a plugin from it!
	... 10 more
```

I assume it's just the API and not the plugin itself - how would I use this? (Sorry if this is a dumb question)

**Comment by D3v1s0m** [comment]
> Also I'm confused about the code you sent, with the new commit you should be able to just do this
> 
> ```java
> return NpcApiProvider.get().getNpcSerializerRegistry().getSerializer(YamlConfiguration.class).serialize(entry).saveToString();
> ```

Why are we exposing a method to serialize a whole npc entry? isn't it safer to expose just the property serialization we already have to property registry?

We also have this method to register dummy properties with classes that have a serializer registered.
![image](https://github.com/user-attachments/assets/b6fd3b36-4a7e-4122-acbe-d88f42bc90d9)

So if a plugin registers a new property, they can fetch the serialized value for that specific property instead of having to serialize the whole npc entry and then find the property in that serialized object

**Comment by Pyrbu** [comment]
> Why are we exposing a method to serialize a whole NPC entry? isn't it safer to expose just the property serialization we already have to property registry?

It isn't safer at all because if we decided to change the code that saves the NPC then his plugin would no longer save the NPCs correctly.

> We also have this method to register dummy properties with classes that have a serializer registered. ![image](https://private-user-images.githubusercontent.com/49519439/398519791-b6fd3b36-4a7e-4122-acbe-d88f42bc90d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMzg0ODIsIm5iZiI6MTczNTEzODE4MiwicGF0aCI6Ii80OTUxOTQzOS8zOTg1MTk3OTEtYjZmZDNiMzYtNGE3ZS00MTIyLWFjYmUtZDg4ZjQyYmM5MGQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDE0NDk0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVlMThkOWU3ZDlmYjI5YjEzNDZmM2E0ZmRkNTQ0YmVlNDc2MjgxOTI4MjhkODJmMzVjMGU1NTUyODY0OGUxMTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fLbOqPH-lMg6K1Ti9DUOTKYQciGZf75PrSMLJO7h3F8)
> So if a plugin registers a new property, they can fetch the serialized value for that specific property instead of having to serialize the whole npc entry and then find the property in that serialized object

That wouldn't work for his use case because he is trying to save the NPC himself as a "blueprint" to create multiple copies

**Comment by tastybento** [comment]
[I've done the code ](https://github.com/BentoBoxWorld/BentoBox/blob/ce84d66b91b763442f5cfcf0024c9cfd194abb45/src/main/java/world/bentobox/bentobox/hooks/ZNPCsPlusHook.java#L35)but it'd be good to test it. I'm not familiar with Gradle and hit some issues trying to compile your code (I know Maven, but Gradle is like magic to me!) and the API snapshot doesn't see to be usable. Is there a link to a SNAPSHOT of your ZNPCs plugin I may use? I'd like to confirm it all works before doing a BentoBox release.

**Comment by ErV1nTract** [comment]
> I tried to use `znpcsplus-api-2.0.0-20241224.151437-145.jar` but I get this error:

Why didn't anyone tell the guy that the jar file with the API needed to be added to the build files instead of the plugins folder?

**Comment by Pyrbu** [comment]
> [I've done the code ](https://github.com/BentoBoxWorld/BentoBox/blob/ce84d66b91b763442f5cfcf0024c9cfd194abb45/src/main/java/world/bentobox/bentobox/hooks/ZNPCsPlusHook.java#L35)but it'd be good to test it. I'm not familiar with Gradle and hit some issues trying to compile your code (I know Maven, but Gradle is like magic to me!) and the API snapshot doesn't see to be usable. Is there a link to a SNAPSHOT of your ZNPCs plugin I may use? I'd like to confirm it all works before doing a BentoBox release.

I actually typed out a second comment explaining that you needed to add the repository and the dependency to your pom.xml file but I guess that didn't send for some reason.

You can get development builds of the plugin built straight from github from our Jenkins https://ci.pyr.lol/job/ZNPCsPlus/

**Comment by D3v1s0m** [comment]
> It isn't safer at all because if we decided to change the code that saves the NPC then his plugin would no longer save the NPCs correctly.

If the whole NPC is serialized and stored in a blueprint by our plugin, and then we change something in the serialization (like keys), then the blue print will not load. Instead, if they code their own serialization and load the NPC using the API (which shouldn't break anything), the blueprint has a much higher chance to not break. The only serialization methods we need to provide in the api is for properties and actions for which we already have registered serializers so we just need to expose the methods.

**Comment by Pyrbu** [comment]
> If the whole NPC is serialized and stored in a blueprint by our plugin, and then we change something in the serialization (like keys), then the blue print will not load. Instead, if they code their own serialization and load the NPC using the API (which shouldn't break anything), the blueprint has a much higher chance to not break. The only serialization methods we need to provide in the api is for properties and actions for which we already have registered serializers so we just need to expose the methods.

It will load because we need to support backwards compatibility in the deserialization, like we already do with null checks and default values. If we didn't support backwards compatibility in the deserializer we would need to make a data fixer system.

**Comment by D3v1s0m** [comment]
> It will load because we need to support backwards compatibility in the deserialization, like we already do with null checks and default values. If we didn't support backwards compatibility in the deserializer we would need to make a data fixer system.

That is what I mean. The blueprints would need to be resaved if the serialization is changed, which destroys the entire functionality of blueprints (reuse as many times as you want with different parameters(origin in this cases as far as I can tell))

**Comment by tastybento** [comment]
[I merged in ZNPCsPlus](https://github.com/BentoBoxWorld/BentoBox/pull/2575) support to BentoBox just now. It'll be in the latest snapshot on ci.bentobox.world and I expect I'll do a release this weekend, if possible.

If I may chip in about the conversation above - It’s great to see such a thoughtful discussion around serialization, and I truly appreciate the effort put into this feature. The main use case is that admins will make starter islands using ZNPCs and save their islands for their players to use. From my perspective, the current approach works brilliantly and meets my needs for creating and managing blueprints. That said, I can understand the concerns raised, especially regarding potential issues with compatibility and future-proofing.

For now, the current implementation seems robust, especially since there’s a clear intention to maintain backward compatibility in the deserialization process through null checks and default values. This gives me confidence that changes in serialization won’t immediately break existing blueprints. If the serialization structure were to evolve significantly in the future, a versioning system might help manage these changes. By including a version number in the serialized data, older blueprints could still be supported, and any required transformations (like through a data fixer) could be applied conditionally. I doubt there will be major changes requiring this though.

Another possible approach, if compatibility becomes a greater concern, could involve ignoring unused or unexpected YAML values during deserialization. This way, even if new properties or keys are introduced in the future, they wouldn’t disrupt the loading of existing blueprints. This method aligns well with a philosophy of resilience and backward compatibility.

For my use case, having a simple, reliable way to serialize the whole NPC is invaluable. However, if it ever becomes necessary to shift to more granular property-based serialization, I’d be happy to adapt and use the API as recommended. For now, I think this feature has been implemented thoughtfully and should serve the community well.

Thank you again for your hard work—it’s greatly appreciated!

**Comment by tastybento** [comment]
@Pyrbu FYI: I released 3.2.0 of BentoBox just now https://github.com/BentoBoxWorld/BentoBox/releases/tag/3.2.0
